### ISSUE TITLE
--watch doesn't work in latest (1.4.15) webpack on osx yosemite 10.10

### ISSUE THREAD

**Issue by neverov** [webpack/webpack#675]
I have a single js file `entry.js` with `document.write("foo")` line and run `webpack --watch entry.js bundle.js`. Webpack starts in watch mode, then I make some changes to a file and can't see the update to the bundle.

**Comment by sokra** [comment]
You can try the NewWatchingPlugin:

``` js
// webpack.config.js
var webpack = require("webpack");
module.exports = {
  plugins: [
    new webpack.NewWatchingPlugin()
  ]
};
```

**Comment by neverov** [comment]
Nope, doesn't work either( 
Maybe I can somehow debug this issue?

**Comment by sokra** [comment]
You can run the tests here: https://github.com/webpack/watchpack

**Comment by neverov** [comment]
hm, all tests passed
ok, I'll try to reinstall everything from scratch and try again

**Comment by sokra** [comment]
So in general the watching works fine... (NewWatchingPlugin)

**Comment by neverov** [comment]
Well, still no idea why this happened, after several restarts and reinstalls of node and npm everything went back to normal. Maybe some hardware issue?

**Comment by amccloud** [comment]
I'm experiencing the same issue on OS X 10.9.5 with webpack 1.5.x.
When i downgrade to webpack 1.4.x everything works fine.

**Comment by andreypopp** [comment]
I second that, webpack 1.5.x doesn't watch files properly on Mac OS X (triggers only on second or third change) while webpack 1.4.x works fine.

**Comment by neverov** [comment]
@amccloud @andreypopp should I reopen this one?

**Comment by mindjuice** [comment]
FYI, this is a bug in OS X.  Try the suggestions listed in this article to fix the problem:

http://feedback.livereload.com/knowledgebase/articles/86239-os-x-fsevents-bug-may-prevent-monitoring-of-certai

Here is a copy of the meat of the article in case that link ever goes away:

>**Fixing it**
>Because we don't know the real cause of the bug, there is no surefire way to resolve it. However, one of the following actions has always helped so far:
>
>* rebooting the computer
>* checking the disk and repairing permissions via Disk Utility
>* adding the folder to Spotlight privacy list (the list of folders to not index), and then removing from it, effectively forcing a reindexing
>* renaming the folder, and then possibly renaming it back
>* re-creating the folder and moving the old contents back into it

**Comment by railsnerd** [comment]
Is it definitely the OSX bug?  The fix list doesn't help

**Comment by IngwiePhoenix** [comment]
1.4 is not the latest.

Ingwie@Ingwies-Macbook-Pro.local ~/W/BIRD3 $ webpack --version
webpack 1.8.9

I am on 10.10 myself and it works. o.o

**Comment by bbirand** [comment]
I'm using webpack 1.9.7 and I also noticed this problem. As noted above, this is likely due to a bug in OS X, and therefore won't be fixed anytime soon..

**Comment by ylg** [comment]
Occurs on 1.9.11 and OS X 10.10.3. Re-cloning the affected directory from git fixes things for a few minutes, then the problem recurs (permanently until another reclone.) No other workaround had effect, e.g., renaming the directory, rebooting. 

mocha --watch, DropBox, and Google Drive are all watching this directory for changes and catching them successfully. Perhaps the mechanism webpack is relying on _is_ broken in OS X, but other tools seem to have worked around it or just not used that particular mechanism?

**Comment by ying22** [comment]
I have the same problem. I figured out a temporary solution. 
I have to move the dir once, e.g:
if your webpack watch dir is at: /Users/you/work/project1/js/src
then:
mv /Users/you/work /Users/you/work123
mv /Users/you/work123 /Users/you/work

It works for a while after the move.

**Comment by sokra** [comment]
@es128

**Comment by es128** [comment]
Seeing a lot of interesting anecdotes here, but unfortunately not much that gives me a clue what's going wrong. In some cases it does look like it might be related to switching versions of node or io.js. As such, `npm rebuild` should always be the first thing to try in these circumstances. 

The other potential solution I'd recommend for anyone suspecting something odd/wrong with their local system is switching watch modes - try using the polling mode (which is the equivalent of what mocha's watch uses).

**Comment by es128** [comment]
@ylg @ying22 what editor are you using when you experience the issue? The description of how it works at first and then loses track of subsequent changes is reminiscent of issues I've seen in the past caused by specific editors employing complex write behaviors.

**Comment by ylg** [comment]
@es128 Occurs for me with Atom 0.210.0, SublimeText 3083, and vim 7.3 (Apple's build). Adding --watch-polling doesn't improve the situation, i.e., neither works once a dir has gone bad:
```shell
% node_modules/webpack/bin/webpack.js --watch
% node_modules/webpack/bin/webpack.js --watch --watch-polling
```

**Comment by ying22** [comment]
I am using sublime 3083. It stopped working after webpack watch and text editor were left running for a few days last time. But I just got back from a weekend trip and it is fine right now.

**Comment by es128** [comment]
@ylg Sounds very strange. If you're willing to spend some time debugging, I'd be interested in the result if you were to use the example script in chokidar's repo, logging out the `raw` event and performing the same sorts of editing actions. 

I wonder if there is something about Dropbox and/or Google Drive contributing to these issues, although that isn't something I've seen reported before.

**Comment by ylg** [comment]
I tried shutting down Dropbox and Google Drive and then making some changes; they were not caught by watch. Of course, it may be that one of those two had already permanently changed the directory, and so I'll try a new clone the next time I'm at a point without a bunch of outstanding code changes.

I tried Miller's example script for chokidar (as shown in the README at https://github.com/paulmillr/chokidar) and it reported "change [filename]" each time I saved a file.

PS If we can confirm @ying22's observation that taking weekend trips may resolve the issue, I'd prefer _no fix_ be applied to the codebase.

**Comment by ying22** [comment]
I don't have dropbox or google drive

**Comment by es128** [comment]
Lol good call @ylg. 

So if chokidar by itself *is* correctly reporting change events under the same circumstances where webpack has trouble, then perhaps there is an issue somewhere upstream of the watch layer. Sounds like it's going to be tough to diagnose and debug though given the inconsistencies you've described.

**Comment by es128** [comment]
Btw this is the example script I was referring to for diagnostic purposes (with the `raw` listener uncommented): https://github.com/paulmillr/chokidar/blob/master/example.js

**Comment by ylg** [comment]
Here's an example of the output from that file running via the same version of iojs after changing a file that a normal webpack run will bundle up, but watching missed, i.e., I had a --watch running at the same time as example.js:

```
change /My/path/file.js
RawEvent: modified /My/path/file.js { path: '/My/path/file.js',
  event: 'modified',
  type: 'file',
  changes: { inode: true, finder: false, access: false, xattrs: false },
  flags: 70656 }
```

One thing of interest, though probably not importance, is that the path capitalization is not the same as reported by Z shell, e.g., chokidar reports /My/path/file.js while Z prints /My/**P**ath/file.js, i.e., only some capitalization is kept.

**Comment by es128** [comment]
@ylg the capitalization issue may be important. Do you have the option to rename those dirs to be lowercase only and see if that has any impact on the issue? Did you specify `path` vs `Path` in the dirs/glob patterns you're watching, or did it just come out of the system with the wrong capitalization on its own?

At least, it seems we have determined at this point that chokidar is NOT missing `change` events and the bug causing the problem you've observed is somewhere else.

**Comment by ying22** [comment]
My top level dir was also capitalized before: "/Users/me/Work". Since the move, I have named it lowercase by accident: "/Users/me/work". And I haven't had any problems since.

**Comment by es128** [comment]
Sounds like there may be some sort of issue with inconsistent capitalization in paths, but would need someone to come up with a clear reproducible case to be able to do anything about it.

**Comment by mikeknoop** [comment]
`webpack --watch` wasn't picking up any changes (but another task `gulp.watch` was).

I renamed my parent directory from

`/Users/mikeknoop/Code/zapier` -> `/Users/mikeknoop/code/zapier`

and now webpatch watching catches changes.

**Comment by akwuh** [comment]
guys... Have the same issue, just run `sudo webpack --watch` all works now. But anyway smth wrong with webpack because sudo is not needed obvoiously

**Comment by bbirand** [comment]
@akwuh Brilliant! This does work.

This issue should **not** be closed, people are still having problems with this..

**Comment by Elyx0** [comment]
I was going crazy since a while with this non-watching issue, please for the love of mankind make a reference to this post somewhere. It saved me. Both `renaming` without uppercases and `sudo` solved it. It doesn't deserve a closed status. Thanks @akwuh & @mikeknoop

**Comment by arshtepe** [comment]
I have this problem on Ubuntu 14.04 Lts, webpack 1.12.11

**Comment by bs85** [comment]
@akwuh @bbirand @arshtepe Just had the problem on Ubuntu 15.10, on a haunch checked if there was a limitation on the amount of files watched as my project is pretty big.

Turns out increasing inotify max watchers fixed it. Sudo seem to be a workaround as this limit doesn't affect superusers, but it's annoying because then the .js files are created as root.

https://github.com/guard/listen/wiki/Increasing-the-amount-of-inotify-watchers

**Comment by frontsideair** [comment]
I also did increase the number of watchers and it was successful for me. I experienced it before but totally forgot about it, even though it has a section in Troubleshooting doc: https://webpack.github.io/docs/troubleshooting.html

**Comment by bbirand** [comment]
@bs85 Thanks for the hint! But I'm on OS X, so not sure how to go about increasing the watchers there.

**Comment by frontsideair** [comment]
@bbirand For OSX, you should check this one: https://webpack.github.io/docs/troubleshooting.html#os-x-fsevents-bug

**Comment by StephanBijzitter** [comment]
**This is most definitely not a bug on OS X**. grunt-contrib-watch has no issues for me, but the watch option on webpack doesn't do anything at all.

**Comment by es128** [comment]
@StephanBijzitter grunt-contrib-watch is based on stat polling. A setting to use polling is available here too if you prefer it that way.

**Comment by BlueVajra** [comment]
My issue is inline with the this bug referenced above. `sudo webpack` doesn't seem to fix it. For those on OSX, if you 'add a new line' does the watch command kick in? For me it does, but just adding a missing semi-colon does not.

**Comment by kiranps** [comment]
@BlueVajra i use linux i have the same problem. have you tried webpack-dev-server, it works for inline changes

**Comment by StephanBijzitter** [comment]
@es128 which, as I said, does not seem to do anything.

**Comment by BlueVajra** [comment]
@kiranps yes webpack-dev-server does lint fine, but has other pipeline issues for us, so it is not ideal

**Comment by keul** [comment]
Seems that the `sudo` trick is OK for running `webpack --watch` but is not working for the webpack-dev-server: I see no reload there.

**Comment by modosc** [comment]
on osx, seeing the same thing.

i noticed this tends to happen after a failed build - if i correct the error `webpack` will build the old version. once i kill of and restart the build is correct.

also, fwiw:
1. adding a newline doesn't correctly trigger the `watch` command
2. tried with `--watch-polling`, same results
3. i verified that `gulp watch` works fine in the same project

**Comment by keul** [comment]
It's probably useful to collect also OSX versions here. For example: I've the issue on OSX 10.10.5
Maybe the issue do not touch El Capitan users?

**Comment by StephanBijzitter** [comment]
@keul using El Capitan, it's most definitely not an issue with the OS.

**Comment by keul** [comment]
@StephanBijzitter thanks for the info, sorry, some old comments above lead me to the wrong path.

I can confirm that the uppercare to lowecase trick worked in my environment: renaming the parent directory of the project from "/Users/keul/Projects" to "/Users/keul/projects" did the trick.
Hope this will not stop working after a while.

**Comment by dslaugh** [comment]
The problem for me was that I needed to update npm. I had an old version and as soon as I updated, it started working.

**Comment by thejoecode** [comment]
Ran into a similar issue, found this and wanted to share my solution.

File changes were being detected, it just appeared that they were not. I ran watch with progress and saw that every save / touch something was happening, just no output.
`webpack --watch --progress`

The issue turned out to be an incorrectly configured context. It would allow first build to run, but watch failed to output 95% of the time. Sometimes it would output one time and then never again. Very hard to diagnose.

```
module.exports = {
context: path.resolve('./modules'),
```

My solution was to remove the context, and put 'modules' in front of the path in each entry. Have not had an issue since.

In short, try the --progress flag if you think your watch is not picking up file system changes. It could just be the rebuild process is failing silently.

**Comment by pfarnach** [comment]
I was having similar issues and they were solved by updating npm.

**Comment by stelund** [comment]
@pfarnach what does "update npm" mean? It is a part of the node installation. Are you upgrading to a new version of node?

**Comment by pfarnach** [comment]
I should be more specific, I updated node from 0.12.2 to 5.7.1 using NVM (Node Version Manager), which subsequently updated my NPM from 2.7.5 to 3.6.0.  (Side note: I highly recommend using NVM)

**Comment by jordizle** [comment]
Webpack uses the new NewWatchingPlugin() by default in the latest version. I had to force the old watch plugin:

`  plugins: [
    new webpack.OldWatchingPlugin()
  ]`

This resolved the issue for me.

**Comment by richpixel** [comment]
Switching to the 'OldWatchingPlugin' worked for me as well... I'm running OSX 10.10.5 (Yosemite)

**Comment by sgarcia-dev** [comment]
Windows 10 x64 here, also had to `new webpack.OldWatchingPlugin()` to get the watch service to work correctly.

**Comment by davewallace** [comment]
Windows 10 x64 here, nothing above works for me so far.

**Comment by user414** [comment]
Tried @jordizle suggestion and didn't work for me on ubuntu 14.04 LTS. This happened after a recent upgrade as it was working before. Might try downgrading some parts and see where this came from.

**Comment by cho-is** [comment]
happening for me too on Ubuntu 14.04.4.

reloads when editing with vim but not with sublime 3

**Comment by amcdnl** [comment]
Seems in Webpack2 the `OldWatchingPlugin` is not there anymore. Anyone got any other hacks?

Heres my config - https://gist.github.com/amcdnl/1fefceb40d57414943e20c1b3976bb18

**Comment by nahumACL** [comment]
same problem here `OldWatchingPlugin` is not longer here https://github.com/webpack/webpack/pull/2604/files :/

**Comment by LasseRafn** [comment]
Any status?

**Comment by mr-nicky** [comment]
I have the same issue. (mac os - el capitan 10.11.6).
Solved by removing brackets "`[]`" from the path.
For example: 
`/Users/username/test` works fine.
`/Users/username/test [0]` doesn't work.

**Comment by decadef20** [comment]
same issue here. --watch-poll saves my day.

**Comment by discordianfish** [comment]
I'm having the same issue on ubuntu 16.04 with 1.13.3

**Comment by SpaceK33z** [comment]
Closing this issue. If you still find a bug with watching, file it in the [watchpack](https://github.com/webpack/watchpack/issues) repo.

Note that there are known issues when your parent directory contains special characters such as brackets or parentheses. See https://github.com/webpack/watchpack/issues/42 for that.

**Comment by Genovo** [comment]
in Ubuntu 16.04
Renaming file  from "folder (2)"  to "folder2"  worked for me.
